### DISCUSSION TITLE
Failed te execute '/usr/bin/mosquitto_pub'

### DISCUSSION THREAD

**Discussion by robinfiddelaers** [SBFspot/SBFspot#684] [Q&A]
Hi, 

I'm running into an issue after migrating my HA from RPi to my NAS.
Had everything configured (Home Assistant / SBFspot / Mosquitto) to read out my SMA-inverters, however I ran into some storage issues & OS-issues on the RPi.

Last week I decided to move the HA to a VM (Debian-12) and followed all directions (https://community.home-assistant.io/t/installing-home-assistant-supervised-using-debian-12/200253)

I restored the backup of my RPi, but of course I have to re-install SBFspot.  I did this yesterday, but I'm running into MQTT-isues, and I don't have an idea how to solve them ...

To be honest, I think I know where the problem is located : HA runs in a docker container, and so does Mosquitto (as it is installed as HA-addon).

SBFspot is not, so it can't reach /usr/bin/mosquitto_pub, simply because it's not there.
But no idea which path I should use in SBFspot.cfg to publish my values to MQTT

Help & suggestions are appreciated !

Down here you can see the output of /usr/local/bin/sbfspot.3/SBFspot -v -finq -nocsv -mqtt
So, there is a connection, there are data, but they won't publish to MQTT ( for reasons I mentioned up here)
```
SBFspot V3.9.7
Yet another tool to read power production of SMA solar inverters
(c) 2012-2022, SBF (https://github.com/SBFspot/SBFspot)
Compiled for Linux (LE) 64 bit with SQLite support
Commandline Args: -v -finq -nocsv -mqtt
Reading config '/usr/local/bin/sbfspot.3/SBFspot.cfg'
Mon Dec  4 11:59:14 2023: INFO: Starting...
sunrise: 08:24
sunset : 16:32
Connecting to 00:80:25:17:BD:F8 (1/10)
Initialising...
SUSyID: 125 - SessionID: 987329711
SMA netID=02
SUSyID: 78 - SN: 2100344218
SUSyID: 78 - SN: 2100327739
BT Signal=66.3%
Logon OK
Local Host Time: 04/12/2023 11:59:17
Plant Time     : 04/12/2023 12:59:16 (+3599 sec)
TZ offset      : 3600 sec - DST: On
Last Time Set  : 03/12/2023 16:25:03
Time was already adjusted in last 7 days
SUSyID: 78 - SN: 2100344218
Device Name:      SN: 2100344218
Device Class:     PV-omvormers
Device Type:      SB 4000TL-20
Software Version: 03.01.05.R
SUSyID: 78 - SN: 2100327739
Device Name:      SN: 2100327739
Device Class:     PV-omvormers
Device Type:      SB 4000TL-20
Software Version: 03.01.05.R
Packet status: 21
Packet status: 21
SUSyID: 78 - SN: 2100344218
Device Status:      OK
SUSyID: 78 - SN: 2100327739
Device Status:      OK
SUSyID: 78 - SN: 2100344218
Device Temperature: 48.7°C
SUSyID: 78 - SN: 2100327739
Device Temperature: 48.5°C
SUSyID: 78 - SN: 2100344218
GridRelay Status:      Gesloten
SUSyID: 78 - SN: 2100327739
GridRelay Status:      Gesloten
SUSyID: 78 - SN: 2100344218
Energy Production:
	EToday: 0.015kWh
	ETotal: 53380.291kWh
	Operation Time: 54298.57h
	Feed-In Time  : 52174.25h
SUSyID: 78 - SN: 2100327739
Energy Production:
	EToday: 0.038kWh
	ETotal: 52861.493kWh
	Operation Time: 54307.83h
	Feed-In Time  : 52202.01h
SUSyID: 78 - SN: 2100344218
DC Spot Data:
	MPPT 1 Pdc:   0.025kW - Udc: 259.44V - Idc:  0.098A
	MPPT 2 Pdc:   0.012kW - Udc: 167.08V - Idc:  0.078A
	Calculated Total Pdc:   0.037kW
AC Spot Data:
	Phase 1 Pac :   0.018kW - Uac: 228.86V - Iac:  0.090A
	Phase 2 Pac :   0.000kW - Uac:   0.00V - Iac:  0.000A
	Phase 3 Pac :   0.000kW - Uac:   0.00V - Iac:  0.000A
	Total Pac   :   0.020kW - Calculated Pac:   0.018kW
	Efficiency  :   48.65%
SUSyID: 78 - SN: 2100327739
DC Spot Data:
	MPPT 1 Pdc:   0.054kW - Udc: 281.89V - Idc:  0.193A
	MPPT 2 Pdc:   0.008kW - Udc: 163.78V - Idc:  0.056A
	Calculated Total Pdc:   0.062kW
AC Spot Data:
	Phase 1 Pac :   0.047kW - Uac: 224.69V - Iac:  0.210A
	Phase 2 Pac :   0.000kW - Uac:   0.00V - Iac:  0.000A
	Phase 3 Pac :   0.000kW - Uac:   0.00V - Iac:  0.000A
	Total Pac   :   0.047kW - Calculated Pac:   0.047kW
	Efficiency  :   75.81%
SUSyID: 78 - SN: 2100344218
Grid Freq. : 49.98Hz
SUSyID: 78 - SN: 2100327739
Grid Freq. : 49.98Hz
SUSyID: 78 - SN: 2100344218
Current Inverter Time: 04/12/2023 12:59:16
Inverter Wake-Up Time: 04/12/2023 10:56:22
Inverter Sleep Time  : 04/12/2023 12:59:18
SUSyID: 78 - SN: 2100327739
Current Inverter Time: 04/12/2023 12:59:15
Inverter Wake-Up Time: 04/12/2023 10:31:09
Inverter Sleep Time  : 04/12/2023 12:59:17
MQTT: Publishing (sbfspot_{serial}) "Timestamp": "04/12/2023 11:59:22","SunRise": "04/12/2023 08:24:00","SunSet": "04/12/2023 16:32:00","InvSerial": 2100344218,"InvName": "SN: 2100344218","InvTime": "04/12/2023 12:59:16","InvStatus": "OK","InvTemperature": 48.720,"InvGridRelay": "Gesloten","EToday": 0.015,"ETotal": 53380.291,"PACTot": 20.000,"UDC1": 259.440,"UDC2": 167.080,"IDC1": 0.098,"IDC2": 0.078,"PDC1": 25.000,"PDC2": 12.000
sh: 1: **/usr/bin/mosquitto_pub: not found**
**MQTT: Failed te execute '/usr/bin/mosquitto_pub' mosquitto client installed?**
MQTT: Publishing (sbfspot_{serial}) "Timestamp": "04/12/2023 11:59:22","SunRise": "04/12/2023 08:24:00","SunSet": "04/12/2023 16:32:00","InvSerial": 2100327739,"InvName": "SN: 2100327739","InvTime": "04/12/2023 12:59:15","InvStatus": "OK","InvTemperature": 48.530,"InvGridRelay": "Gesloten","EToday": 0.038,"ETotal": 52861.493,"PACTot": 47.000,"UDC1": 281.890,"UDC2": 163.780,"IDC1": 0.193,"IDC2": 0.056,"PDC1": 54.000,"PDC2": 8.000
sh: 1: **/usr/bin/mosquitto_pub: not found**
**MQTT: Failed te execute '/usr/bin/mosquitto_pub' mosquitto client installed?
Error 32512 while publishing to MQTT Broker**
********************
* ArchiveDayData() *
********************
startTime: 04/12/2023 00:00:00
**********************
* ArchiveMonthData() *
**********************
startTime: 01/12/2023 12:00:00
Reading events: 2023-Dec-01
Mon Dec  4 11:59:24 2023: INFO: Done.
```
Thanks for some help or suggestions  !
regards,
Robin

**Comment by sbf-** [comment]
Run SBFspot in a second container and let the two communicate with each other? However, I have no experience with this, but maybe [this](https://superuser.com/questions/1676124/how-to-run-mqtt-broker-and-clients-in-different-docker-containers) might be a good starting point.

**Reply by robinfiddelaers** [reply]
Hi, thanks for your input.  I'm not much of an expert (rather new in VM & docker stuff) but have looked into it.
I can be wrong, but I don't think this would help me out. If I would run SBFspotin a different container, how would it be possible for SBFspot to reach the /usr/bin/ path on the mosquitto container to execute /usr/bin/mosquitto_pub ?
What I think for now, is that the 2 applications need topee on the same file system, so mounting the container FS  on the FS where SBFspot is the only solution I presume ? Or maybe to find a way to install a second MQTT-publisher on the SBFspot FS and let it use that to publish the values, so Mosquitto can pick them up in Home Assistant.  But isn't that a bit of 'killing a mosquitto with a sledgehammer' ? :-/

**Reply by sbf-** [reply]
> But isn't that a bit of 'killing a mosquitto with a sledgehammer' ?

Why? SBFspot needs mosquitto_pub to be able to publish its data to the MQTT broker.
There is a [docker container](https://github.com/SBFspot/SBFspot/tree/master/docker) with MQTT support available.
